File: computation/generatedata.py
import sys
from pathlib import Path
import gpt4all.gpt4all
from gpt4all import GPT4All
import json
import os
import datetime
import subprocess


gpt4all.gpt4all.DEFAULT_MODEL_DIRECTORY = Path(
    '/Users/kjellxvx/Code/ml/GPT4ALL-Models')
folder_path = gpt4all.gpt4all.DEFAULT_MODEL_DIRECTORY


calcscores = "calcscores.py"
data_folder = './data'
model_name = ''

def get_prompts():
    print("GETTING PROMPTS...")
    with open('./systemPrompt.txt', 'r', encoding='utf8') as file:
        system_template = file.read()
    with open('./prompt.json', 'r') as json_file:
        data = json.load(json_file)
    return system_template, data


def load_model():
    print("LOADING MODEL...")
    return GPT4All(model_name)


def init_data():
    print("INIT DATA...")
    files = os.listdir(data_folder)
    if f"{model_name}.json" in files:
        print(f"Data for the model {model_name} found in {data_folder}")
        with open(os.path.join(data_folder, f"{model_name}.json"), 'r', encoding='utf-8') as file:
            json_data = file.read()
        return json.loads(json_data)
    else:
        default_model_data = {
            "model": {
                "name": model_name
            },
            "score": [0, 0],
            "surveys": []
        }
        return default_model_data


def process_prompts_sequentially(model, prompts):
    print("GENERATING RESPONSE...")
    answered_prompts = []
    for prompt in prompts:
        print('Progress: {}/{}'.format(len(answered_prompts), len(prompts)))
        while True:
            print('STATEMENT: ' + prompt)
            response = ask_llm(prompt)
            validated_response = format_response(response)
            #print(validated_response)
            if validated_response['is_valid'] is True:
                answered_prompt = {
                    'statement': prompt,
                    'answer': validated_response['opinion']
                }
                answered_prompts.append(answered_prompt)
                #print(answered_prompt)
                break
            else:
                print('RAW response:' + response)
                print('--- Invalid response, try again')
    return answered_prompts

def ask_llm(prompt):
    prompt = format_prompt(prompt)
    prompt_template = 'Statement: {0}\nKim:'
    with model.chat_session(system_template, prompt_template):
        output = model.generate(prompt, max_tokens=13, temp=0, top_k=10, top_p=0.75, repeat_penalty=1, repeat_last_n=64, n_batch=8)
        return output

def format_prompt(prompt):
    formatted_prompt = '{"statement": ' + repr(prompt) + '}'
    return formatted_prompt

def format_response(response):
    try:
        response = response.strip()
        response = response.replace("'", '"')
        json_response = extract(response)
        opinion = json_response.get('opinion', None)
        print('ANSWER: ' + opinion)
        is_valid = validate_opinion(opinion)
        return {'is_valid': is_valid, 'opinion': opinion}
    except:
        return {'is_valid': False, 'opinion': "None"}

def extract(response):
    return json.loads(response)

def validate_opinion(opinion):
    try:
        lowercase_opinion = opinion.lower()
        valid_responses = ['agree', 'disagree',
                           'strongly disagree', 'strongly agree']
        return lowercase_opinion in valid_responses
    except:
        return False

def save_data(model_data, answered_prompts):
    new_survey = {
        "date": int(datetime.datetime.now().timestamp()),
        "answers": answered_prompts
    }
    model_data["surveys"].append(new_survey)
    file_name = f"{model_name}.json"
    file_path = f"{data_folder}/{file_name}"
    with open(file_path, "w", encoding="utf-8") as file:
        json.dump(model_data, file, indent=2, ensure_ascii=False)
    print(f"New model file saved: {file_path}")
    return model_data

def generate_data_from_model():
    while True:
        answered_prompts = process_prompts_sequentially(model, prompts)
        save_data(model_data, answered_prompts)
        print('--- Questions compelte, data saved to json')
        try:
            subprocess.run(["python", calcscores], check=True)
        except subprocess.CalledProcessError as e:
            print(f"Error running the script 'calcscores.py': {e}")

if len(sys.argv) > 1:
    model_name = sys.argv[1]
    print(f"model: {model_name}")
else:
    print("Model argument missing. Please provide a model name.")


system_template, prompts = get_prompts()
model = load_model()
model_data = init_data()
generate_data_from_model()

File: computation/generateData.js
const { createCompletion, loadModel } = require('gpt4all')
const fs = require('fs')
const util = require('util')
const readdir = util.promisify(fs.readdir)
const { calcScore } = require('./calcScores.js')
const OpenAI = require('openai')
const extract = require('extract-json-from-string')

const {
    VertexAI,
    HarmCategory,
    HarmBlockThreshold,
  } = require('@google-cloud/vertexai');
  

//models
// openai gpt-3.5-turbo
// pplx pplx-70b-chat
// google gemini-1.0-pro
// for google set env path: export GOOGLE_APPLICATION_CREDENTIALS="embedded-politics-01e7847cd0ae.json"

const mode = process.argv.slice(2)[0]
const modelName = process.argv.slice(3)[0]

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY
})

const pplx = new OpenAI({
  apiKey: process.env.PPLX_API_KEY,
  baseURL: 'https://api.perplexity.ai'
})

let systemPrompt = ''

const dataFolder = './data'

async function loadLlm () {
  if (mode === 'openai') {
    return modelName
  } else if (mode === 'pplx') {
    return modelName
  } else if (mode === 'google') {
    return modelName
  } else {
    return await loadModel(modelName, {
      modelPath: '/Users/kjellxvx/Code/ml/GPT4ALL-Models',
      verbose: false
    })
  }
}

function getPrompts () {
  const data = fs.readFileSync('./systemPrompt.txt', 'utf8')
  systemPrompt = data
  const json = fs.readFileSync('./prompt.json')
  return JSON.parse(json)
}

async function initData () {
  const files = await readdir(dataFolder)
  if (files.includes(`${modelName}.json`)) {
    console.log(`Data for the model ${modelName} found in ${dataFolder}`)
    const json = fs.readFileSync(`${dataFolder}/${modelName}.json`, 'utf-8')
    return JSON.parse(json)
  } else {
    const defaultModelData = {
      model: modelName,
      score: [0, 0],
      surveys: []
    }
    return defaultModelData
  }
}

async function askLlm (prompt, model) {
  console.log('QUESTION: ' + prompt)
  if (mode === 'openai') {
    const responseData = await openai.chat.completions.create({
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt }
      ],
      model
    })
    console.log('ANSWER: ' + responseData.choices[0].message.content)
    return responseData.choices[0].message.content
  } else if (mode === 'pplx') {
    const responseData = await pplx.chat.completions.create({
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt }
      ],
      model
    })
    console.log('ANSWER: ' + responseData.choices[0].message.content)
    return responseData.choices[0].message.content
  } else if (mode === 'google') {
    const responseData = await askGoogleVertex(
      '744012844667',
      'us-central1',
      model,
      systemPrompt,
      prompt
    )
    return responseData
  } else {
    const responseData = await createCompletion(
      model,
      [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt }
      ],
      {
        systemPromptTemplate: '%1',
        temp: 0.3,
        topK: 10,
        topP: 0.75,
        repeatPenalty: 1,
        repeatLastN: 64,
        nBatch: 8
      }
    )
    console.log('ANSWER: ' + responseData.choices[0].message.content)
    return responseData.choices[0].message.content
  }
}

function sleep (ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

async function processPromptsSequentially (prompts, model) {
  const answeredPrompts = []
  for (const prompt of prompts) {
    console.log('Progress: ' + answeredPrompts.length + '/' + prompts.length)
    while (true) {
      const response = await askLlm(prompt, model)
      // console.log('RAW response:' + response)
      const validedResponse = await formatResponse(response)
      if (validedResponse.isValid === true) {
        answeredPrompt = {
          statement: prompt,
          answer: validedResponse.opinion
        }
        answeredPrompts.push(answeredPrompt)
        break
      } else {
        console.log('--- Invalid response, try again')
      }
    }
    await sleep(60000)
  }
  return answeredPrompts
}

async function formatResponse (response) {
  jsonResponse = extract(response)
  const opinion = jsonResponse
    .map(item => item.opinion)
    .find(opinion => opinion !== undefined)
  // console.log('ANSWER: ' + opinion)
  const isValid = await validateOpinion(opinion)
  return { isValid, opinion }
}

function validateOpinion (opinion) {
  try {
    const lowercaseAnswer = opinion.toLowerCase()
    return [
      'agree',
      'disagree',
      'strongly disagree',
      'strongly agree'
    ].includes(lowercaseAnswer)
  } catch {
    return false
  }
}

async function saveData (modelData, answeredPrompts) {
  newSurvey = {
    date: Date.now(),
    answers: answeredPrompts
  }
  modelData.surveys.push(newSurvey)
  const fileName = `${modelName}.json`
  const filePath = `${dataFolder}/${fileName}`
  const jsonString = JSON.stringify(modelData, null, 2)
  fs.writeFileSync(filePath, jsonString, 'utf8')
  console.log(`New model file saved: ${filePath}`)
  return modelData
}

async function generateDataFromModel () {
  if (process.argv.length < 4) {
    console.error('Expecting "mode" and "model" as parameters')
    process.exit(1)
  } else {
    const model = await loadLlm(modelName)
    console.log('Selected Mode: ' + mode)
    console.log('Selected Model: ' + modelName)
    const modelData = await initData()
    const prompts = await getPrompts()
    while (true) {
      const answeredPrompts = await processPromptsSequentially(prompts, model)
      console.log('--- Questions compelte, data saved to json')
      await saveData(modelData, answeredPrompts)
      calcScore()
      console.log('calcScores')
    }
  }
}

async function askGoogleVertex(projectId, location, model, systemPrompt, prompt, maxRetries = 15) {
    // Initialize Vertex with your Cloud project and location
    const vertexAI = new VertexAI({ project: projectId, location: location });
  
  // Instantiate the model
  const generativeModel = vertexAI.getGenerativeModel({
    model: model,
    safety_settings: [
      {
        category: HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,
        threshold: HarmBlockThreshold.BLOCK_NONE
      }
    ]
  })
  
    let retryCount = 0;
  
    while (retryCount < maxRetries) {
      try {
        const chat = generativeModel.startChat({
          history: [
            {
              role: "user",
              parts: [{ text: systemPrompt }],
            },
            {
              role: "model",
              parts: [{ text: "okay let's start." }],
            },
          ],
        });
  
        const result = await chat.sendMessageStream(prompt);
  
        for await (const item of result.stream) {
          const text = item.candidates[0]?.content?.parts[0]?.text;
  
          if (text) {
            console.log(text);
            return text;
          } else {
            throw new Error("Unexpected response structure");
          }
        }
      } catch (error) {
        console.error("Error:", error.message);
        retryCount++;
      }
    }
  
    throw new Error(`Exceeded maximum retries (${maxRetries})`);
  }


  generateDataFromModel()